src/lib/reasoners/uf.ml

```diff
@@ -999,100 +999,206 @@ let assign_next env =
+    | Some abstract -> abstract
+    | None ->
+      let abstract = Id.Namespace.Abstract.fresh () |> Hstring.make in
+      Hashtbl.add abstracts_cache r abstract;
+      abstract
+
+  let clear () =
+    Hashtbl.clear arrays_cache;
+    Hashtbl.clear records_cache;
+    Hashtbl.clear abstracts_cache
+end
+
+let is_forbidden_symbol f ty =
   (* Keep record constructors because models.ml expects them to be there *)
-  if (X.is_solvable_theory_symbol f ty
-      && not (Shostak.Records.is_mine_symb f ty))
-  || E.is_internal_name t || E.is_internal_skolem t
-  || E.equal t E.vrai || E.equal t E.faux
+  (X.is_solvable_theory_symbol f ty
+   && not (Shostak.Records.is_mine_symb f ty))
   || is_optimization_op f
   || Sy.is_internal f
+
+let compute_concrete_model_of_val env t ((mdl, mrepr) as acc) unbounded =
+  let { E.f; xs; ty; _ } = E.term_view t in
+  if is_forbidden_symbol f ty
+  || E.is_internal_name t || E.is_internal_skolem t
+  || E.equal t E.vrai || E.equal t E.faux
   then
     acc
   else
-    let xs, tys, mrepr =
-      List.fold_left
-        (fun (xs, tys, mrepr) x ->
-           let rep_x, mrepr = model_repr_of_term x env mrepr None in
-           assert (is_const_term rep_x);
-           (x, rep_x)::xs,
-           (E.type_info x)::tys,
-           mrepr
-        ) ([],[], mrepr) (List.rev xs)
-    in
-    let rep, mrepr = model_repr_of_term t env mrepr unbounded in
-    assert (is_a_good_model_value rep);
-    assert (is_const_term rep);
-    match f, xs, ty with
-    | Sy.Op Sy.Set, _, _ -> acc
-
-    | Sy.Op Sy.Get, [(_,(a,_));((_,(i,_)) as e)], _ ->
-      begin
-        match X.term_extract a with
-        | Some ta, true ->
-          let { E.f = f_ta; xs=xs_ta; _ } = E.term_view ta in
-          assert (xs_ta == []);
-          fprofs,
-          cprofs,
-          ModelMap.add (f_ta,[X.type_info i], ty) ([e], rep) carrays,
-          mrepr
-
-        | _ -> assert false
-      end
+    begin
+      let arg_vals, arg_tys, mrepr =
+        List.fold_left
+          (fun (arg_vals, arg_tys, mrepr) arg ->
+             let rep_arg, mrepr = model_repr_of_term arg env mrepr unbounded in
+             rep_arg :: arg_vals,
+             (Expr.type_info arg) :: arg_tys,
+             mrepr
+          )
+          ([], [], mrepr) (List.rev xs)
+      in
+      let ret_rep, mrepr = model_repr_of_term t env mrepr unbounded in
+      match f, arg_vals, ty with
+      | Sy.Name _, [], Ty.Tfarray _ ->
+        begin
+          match Hashtbl.find_opt Cache.arrays_cache t with
+          | Some _ -> acc
+          | None ->
+            (* We have to add an abstract array in case there is no
+               constraint on its values. *)
+            Hashtbl.add Cache.arrays_cache t (Hashtbl.create 17);
+            acc
+        end
+
+      | Sy.Op Sy.Set, _, _ ->
+        (* As arrays are immutable, the result of the set operator is
+           never a user-defined array and has to be ignored. *)
```

Comment:
This is not true. The result of the set operator can be equal to a user-defined array, as in #929 (in fact this may or may not be the source of #929)

Reply:
My comment was misleading. In fact `Set` symbols occurs here by accident because `X.is_my_symbol` isn't the appropriate function to filter terms here. I just removed the comment and I'll do another PR in which a create a better predicate for this use case.